src/lib.rs

```diff
@@ -1,407 +1,316 @@
-            fatal_error("Cannot send bootstrap data to worker pool".to_string());
-        }
-
-        // Wait for the worker pool to be fully bootstraped before moving on.
-        //
-        // It's really important to NOT start the web server before the workers are ready.
-        // Our Kubernetes deployment exposes a readiness probe that relies on the web server
-        // to be listening. The API server will start hitting the policy server as soon as the
-        // readiness probe marks the instance as ready.
-        // We don't want Kubernetes API server to send admission reviews before ALL the workers
-        // are ready.
-        loop {
-            match worker_pool_bootstrap_res_rx.try_recv() {
-                Ok(res) => match res {
-                    Ok(_) => break,
-                    Err(e) => fatal_error(e.to_string()),
-                },
-                Err(oneshot::error::TryRecvError::Empty) => {
-                    // the channel is empty, keep waiting
-                }
-                _ => {
-                    fatal_error("Cannot receive worker pool bootstrap result".to_string());
-                    return;
-                }
-            }
-        }
-        info!(status = "done", "worker pool bootstrap");
-        memory_usage(config.pool_size);
-
-        // All is good, we can start listening for incoming requests through the
-        // web server
-        server::run_server(&config.addr, config.tls_config, api_tx).await;
-
-        // The evaluation is done, we can shutdown the tokio task that is running
-        // the CallbackHandler
-        if callback_handler_shutdown_channel_tx.send(()).is_err() {
-            error!("Cannot shut down the CallbackHandler task");
-        } else if let Err(e) = callback_handle.await {
-            error!(
-                error = e.to_string().as_str(),
-                "Error waiting for the CallbackHandler task"
-            );
-        }
-    });
 
-    if let Err(e) = wasm_thread.join() {
-        fatal_error(format!("error while waiting for worker threads: {e:?}"));
-    };
+        // Stop the CallbackHandler
+        self.callback_handler_shutdown_channel_tx
+            .send(())
+            .expect("Cannot send shutdown signal to CallbackHandler");
 
-    Ok(())
-}
+        // Wait for the CallbackHandler to exit
+        callback_handler
+            .await
+            .expect("Cannot wait for CallbackHandler to exit");
 
-fn memory_usage(pool_size: usize) {
-    let process = match procfs::process::Process::myself() {
-        Ok(p) => p,
-        Err(e) => {
-            warn!(error =? e, "cannot access process stats");
-            return;
-        }
-    };
-    let mem_stats = match process.statm() {
-        Ok(s) => s,
-        Err(e) => {
-            warn!(error =? e, "cannot access process memory stats");
-            return;
-        }
-    };
+        shutdown_tracer_provider();
```

Comment:
we are calling `shutdown_tracer_provider` here in api_server.run(), but also in main(). Should we only call it in main()?